// @shopify/react-native-skia/lib/module/sksg/nodes/colorFilters.js
"worklet";

import { enumKey } from "../../dom/nodes";
import { BlendMode } from "../../skia/types";
export const makeBlendColorFilter = (ctx, props) => {
  const {
    mode
  } = props;
  const color = ctx.Skia.Color(props.color);
  const cf = ctx.Skia.ColorFilter.MakeBlend(color, BlendMode[enumKey(mode)]);
  return cf;
};
export const makeSRGBToLinearGammaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeSRGBToLinearGamma();
  return cf;
};
export const makeLinearToSRGBGammaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeLinearToSRGBGamma();
  return cf;
};
export const declareLerpColorFilter = (ctx, props) => {
  const {
    t
  } = props;
  const second = ctx.colorFilters.pop();
  const first = ctx.colorFilters.pop();
  if (!first || !second) {
    throw new Error("LerpColorFilterNode: missing two color filters as children");
  }
  const cf = ctx.Skia.ColorFilter.MakeLerp(t, first, second);
  ctx.colorFilters.push(cf);
};
export const makeMatrixColorFilter = (ctx, props) => {
  const {
    matrix
  } = props;
  const cf = ctx.Skia.ColorFilter.MakeMatrix(matrix);
  return cf;
};
export const makeLumaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeLumaColorFilter();
  return cf;
};
