// @shopify/react-native-skia/lib/module/sksg/nodes/pathEffects.js
"worklet";

import { composeDeclarations, enumKey, processPath } from "../../dom/nodes";
import { Path1DEffectStyle } from "../../skia/types";
export const makeDiscretePathEffect = (ctx, props) => {
  const {
    length,
    deviation,
    seed
  } = props;
  return ctx.Skia.PathEffect.MakeDiscrete(length, deviation, seed);
};
export const makePath2DPathEffect = (ctx, props) => {
  const {
    matrix
  } = props;
  const path = processPath(ctx.Skia, props.path);
  const pe = ctx.Skia.PathEffect.MakePath2D(matrix, path);
  if (pe === null) {
    throw new Error("Path2DPathEffect: invalid path");
  }
  return pe;
};
export const makeDashPathEffect = (ctx, props) => {
  const {
    intervals,
    phase
  } = props;
  const pe = ctx.Skia.PathEffect.MakeDash(intervals, phase);
  return pe;
};
export const makeCornerPathEffect = (ctx, props) => {
  const {
    r
  } = props;
  const pe = ctx.Skia.PathEffect.MakeCorner(r);
  if (pe === null) {
    throw new Error("CornerPathEffect: couldn't create path effect");
  }
  return pe;
};
export const declareSumPathEffect = ctx => {
  // Note: decorateChildren functionality needs to be handled differently
  const pes = ctx.pathEffects.popAll();
  const pe = composeDeclarations(pes, ctx.Skia.PathEffect.MakeSum.bind(ctx.Skia.PathEffect));
  ctx.pathEffects.push(pe);
};
export const makeLine2DPathEffect = (ctx, props) => {
  const {
    width,
    matrix
  } = props;
  const pe = ctx.Skia.PathEffect.MakeLine2D(width, matrix);
  if (pe === null) {
    throw new Error("Line2DPathEffect: could not create path effect");
  }
  return pe;
};
export const makePath1DPathEffect = (ctx, props) => {
  const {
    advance,
    phase,
    style
  } = props;
  const path = processPath(ctx.Skia, props.path);
  const pe = ctx.Skia.PathEffect.MakePath1D(path, advance, phase, Path1DEffectStyle[enumKey(style)]);
  if (pe === null) {
    throw new Error("Path1DPathEffect: could not create path effect");
  }
  return pe;
};
