// @shopify/react-native-skia/lib/commonjs/sksg/nodes/colorFilters.js
"use strict";
"worklet";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeSRGBToLinearGammaColorFilter = exports.makeMatrixColorFilter = exports.makeLumaColorFilter = exports.makeLinearToSRGBGammaColorFilter = exports.makeBlendColorFilter = exports.declareLerpColorFilter = void 0;
var _nodes = require("../../dom/nodes");
var _types = require("../../skia/types");
const makeBlendColorFilter = (ctx, props) => {
  const {
    mode
  } = props;
  const color = ctx.Skia.Color(props.color);
  const cf = ctx.Skia.ColorFilter.MakeBlend(color, _types.BlendMode[(0, _nodes.enumKey)(mode)]);
  return cf;
};
exports.makeBlendColorFilter = makeBlendColorFilter;
const makeSRGBToLinearGammaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeSRGBToLinearGamma();
  return cf;
};
exports.makeSRGBToLinearGammaColorFilter = makeSRGBToLinearGammaColorFilter;
const makeLinearToSRGBGammaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeLinearToSRGBGamma();
  return cf;
};
exports.makeLinearToSRGBGammaColorFilter = makeLinearToSRGBGammaColorFilter;
const declareLerpColorFilter = (ctx, props) => {
  const {
    t
  } = props;
  const second = ctx.colorFilters.pop();
  const first = ctx.colorFilters.pop();
  if (!first || !second) {
    throw new Error("LerpColorFilterNode: missing two color filters as children");
  }
  const cf = ctx.Skia.ColorFilter.MakeLerp(t, first, second);
  ctx.colorFilters.push(cf);
};
exports.declareLerpColorFilter = declareLerpColorFilter;
const makeMatrixColorFilter = (ctx, props) => {
  const {
    matrix
  } = props;
  const cf = ctx.Skia.ColorFilter.MakeMatrix(matrix);
  return cf;
};
exports.makeMatrixColorFilter = makeMatrixColorFilter;
const makeLumaColorFilter = ctx => {
  const cf = ctx.Skia.ColorFilter.MakeLumaColorFilter();
  return cf;
};
exports.makeLumaColorFilter = makeLumaColorFilter;
